Fix storyboards with no-op alpha operations causing extended drawable lifetimes

peppy opened the pull request:
Test with https://osu.ppy.sh/beatmapsets/139525#osu/348550. See https://github.com/ppy/osu/issues/32453#issuecomment-2742562780 (and inline comments) for explanation.

Closes https://github.com/ppy/osu/issues/32453.

peppy committed:
Fix storyboards with no-op alpha operations causing extended drawable lifetimes

Test with https://osu.ppy.sh/beatmapsets/139525#osu/348550. See
https://github.com/ppy/osu/issues/32453#issuecomment-2742562780 (and
inline comments) for explanation.

Closes https://github.com/ppy/osu/issues/32453.

smoogipoo commented:
I'm a little bit wary that we don't have an evolving set of test cases to cover all these unanticipated behaviours we're seeing...

smoogipoo commented:
Well, at least tests will show _a_ failure, not quite the one I'd be expecting but it is what it is I guess 😓

peppy committed:
Fix test failures due to `StoryboarVideo` having a weird initialisation flow

peppy committed:
Add test coverage of previous failcase

bdach commented:
> The test failures are due to videos' fades being handled in a super non-conventional way

Do you mean this?

https://github.com/ppy/osu/blob/c82eaafe98d96b9f49a4a7f168ef5c484e67d76f/osu.Game/Storyboards/StoryboardVideo.cs#L15-L17

If that is the case, then isn't this fix going to potentially break sync of videos with user transforms specified? Specifically, because of

https://github.com/ppy/osu/blob/17ca29c2c6e0e5f141f5393923a511cfa2d61437/osu.Game/Storyboards/Drawables/DrawableStoryboardVideo.cs#L63-L71

`StartTime` - which is being modified in this PR - is used to initialise playback position. Your fix _will_ work for videos that do not have any alpha transforms specified on the video, but if my reading is correct, then if the video *has* alpha transforms specified, then the playback position of the video will be changed in this PR from its original start time to the start time of the first non-no-op alpha transform, which could potentially desync issues, couldn't it?

Related: https://github.com/ppy/osu/pull/27967#discussion_r1576302734 - I guess nobody has raised an example of a storyboard with videos with alpha transforms on, so maybe an irrelevant concern for the time being. That said this is all starting to feel really inscrutable now.

peppy committed:
Add failing test coverage of regressing video scenario

peppy committed:
Fix video offset start time regression with `StartTime` calculation changes

peppy commented:
Well noticed. I've added test coverage and fixed this. There might still be some unknown incorrect cases of things like users specifying negative start times for video transforms, but generally this should be quite rare / non-existent since stable doesn't even allow applying transforms to videos in the in-game editor.

bdach reviewed (approved):
performance looks improved, other storyboards look maybe not broken, I guess put this in the "try and see whether this is going to need to get insta-reverted or fixed up in post" column
